File: Misc/Python/renderer1.py

'''
    Automatically scale a treeview font size and height of cell. To scale the height of the cell,
add an extra column of height values and bind that to the column. The other properties can be set
as usual.

    Tested with Ubuntu14.04, GTK3.10 and Python2.7

    C. Eric Cashon

'''

from gi.repository import Gtk

class MainWindow(Gtk.Window):

    def __init__(self):
        Gtk.Window.__init__(self, title="Scale Text")

        self.set_default_size(200, 200)
        self.set_position(Gtk.WindowPosition.CENTER)
        self.connect("size_allocate", self.size_changed)

        #The second value is the starting height of the cell.
        self.liststore = Gtk.ListStore(str, int)
        self.liststore.append(["Apples", 30])
        self.liststore.append(["Oranges", 30])
        self.liststore.append(["Pears", 30])

        self.treeview = Gtk.TreeView(model=self.liststore)
        self.treeview.set_vexpand(True)
        self.treeview.set_hexpand(True)

        self.renderer = Gtk.CellRendererText()
        self.renderer.set_property("font", "Arial 10")
        self.renderer.set_property("xalign", 0.5)
        self.renderer.set_property("background", "#ff00ff")

        #Set the column to get the cell height from.
        self.column = Gtk.TreeViewColumn("Fruit", self.renderer, text = 0, height = 1)
        self.column.set_expand(True)

        self.treeview.append_column(self.column)

        self.add(self.treeview)

    def size_changed(self, widget, data):
        width = self.get_allocated_width()
        height = self.get_allocated_height()        
        if(height > 200):
            scale = height / 20
            print(str(height) + " " + str(scale) + " " + str(scale + 20))
            #Set the height of the cells.
            iter1 = self.liststore.get_iter_first()         
            while(iter1 != None):
                self.liststore.set(iter1, 1, scale+20)
                iter1 = self.liststore.iter_next(iter1)  
            #Set the font and background.
            self.renderer.set_property("font", "Arial " + str(scale))
            self.renderer.set_property("background", "#ffff00")
        else:
            self.renderer.set_property("font", "Arial 10")
            self.renderer.set_property("background", "#ff00ff")

win = MainWindow()
win.connect("delete-event", Gtk.main_quit)
win.show_all()
Gtk.main()

File: Misc/Csamples/renderer1.c

/*
    Automatically scale a treeview font size and height of cell. To scale the height of the cell,
add an extra column of height values and bind that to the column. The other properties can be set
as usual.
    This is the C version of renderer1.py. A little code comparison.

    Tested with Ubuntu14.04, GTK3.10

    gcc -Wall renderer1.c -o renderer1 `pkg-config --cflags --libs gtk+-3.0`

    C. Eric Cashon
*/

#include<gtk/gtk.h>

static void size_changed(GtkWidget *widget, GdkRectangle *allocation, gpointer *data)
  {
    gint height=gtk_widget_get_allocated_height(widget);

    if(height>200)
      {
        gint scale=height/20;
        g_print("Height %i, Scale %i, Cell Height %i\n", height, scale, scale+20);
        
        GtkTreeIter iter;
        gboolean valid_iter=FALSE;
        valid_iter=gtk_tree_model_get_iter_first(GTK_TREE_MODEL(data[0]), &iter);
        while(valid_iter)
          {
            gtk_list_store_set(data[0], &iter, 1, scale+20, -1);
            valid_iter=gtk_tree_model_iter_next(GTK_TREE_MODEL(data[0]), &iter);
          }
        
        gchar *string=g_strdup_printf("Arial %i", scale);
        g_object_set(data[1], "font", string, "background", "#ffff00", NULL);  
        g_free(string);
      }
    else
      {
        g_object_set(data[1], "font", "Arial 10", "background", "#ff00ff", NULL);
      }

  }
int main(int argc, char *argv[])
  {
    gtk_init(&argc, &argv);

    GtkWidget *window=gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title(GTK_WINDOW(window), "Scale Text");
    gtk_window_set_position(GTK_WINDOW(window), GTK_WIN_POS_CENTER);
    gtk_window_set_default_size(GTK_WINDOW(window), 200, 200);
    g_signal_connect(window, "destroy", G_CALLBACK(gtk_main_quit), NULL);

    GtkTreeIter iter;
    GtkListStore *store=gtk_list_store_new(2, G_TYPE_STRING, G_TYPE_INT);
    gtk_list_store_append(store, &iter);
    gtk_list_store_set(store, &iter, 0, "Apples",  1, 30, -1);
    gtk_list_store_append(store, &iter);
    gtk_list_store_set(store, &iter, 0, "Oranges", 1, 30, -1);
    gtk_list_store_append(store, &iter);
    gtk_list_store_set(store, &iter, 0, "Pears", 1, 30, -1);

    GtkWidget *tree=gtk_tree_view_new_with_model(GTK_TREE_MODEL(store));
    gtk_tree_view_set_grid_lines(GTK_TREE_VIEW(tree), GTK_TREE_VIEW_GRID_LINES_BOTH);
    gtk_widget_set_hexpand(tree, TRUE);
    gtk_widget_set_vexpand(tree, TRUE);

    GtkCellRenderer *renderer=gtk_cell_renderer_text_new();
    g_object_set(renderer, "xalign", 0.5, "editable", FALSE, "background", "#ff00ff", "font", "Arial 10", NULL);
   
    GtkTreeViewColumn *column=gtk_tree_view_column_new_with_attributes("Fruit", renderer, "text", 0, "height", 1, NULL);
    gtk_tree_view_append_column(GTK_TREE_VIEW(tree), column);    

    gpointer pointers[]={store, renderer};
    g_signal_connect(window, "size_allocate", G_CALLBACK(size_changed), pointers); 

    GtkWidget *scroll=gtk_scrolled_window_new(NULL, NULL);
    gtk_container_add(GTK_CONTAINER(scroll), tree);  

    GtkWidget *grid=gtk_grid_new();
    gtk_grid_attach(GTK_GRID(grid), scroll, 0, 0, 1, 1);
    gtk_container_add(GTK_CONTAINER(window), grid);
   
    gtk_widget_show_all(window);
    gtk_main();
    return 0;   
  }
